[vcpkg] Fixup rpath after building dynamic libraries on linux

Osyotr opened the pull request:
This is required to make packages built with (currently nonexisting) `x64-linux-dynamic` triplet relocatable.
After building a package, we patch binaries so that they have correct rpath values, relative to `lib/` (or `debug/lib/`).
Partially addresses my comment https://github.com/microsoft/vcpkg/pull/19940#issuecomment-1022701293.

Osyotr committed:
Fixup rpath after building dynamic libraries on linux

JackBoosY commented:
We are still discussing about rpath in #19127.

Osyotr commented:
@JackBoosY six months without replies is not really a discussion... Maybe this implementation will reignite the it.

JackBoosY commented:
Fine, I canceled all tests.

dg0yt commented:
> Fine, I canceled all tests.

Why?

Osyotr commented:
@JackBoosY I was just about say that this check is not that useless after all, because some ports support only dynamic builds even on linux.

JackBoosY commented:
/azp run

Osyotr commented:
x64_osx failed but I think it's not because of this PR:
```
duplicate symbol '_METIS_NodeND' in:
    lib/Graclus/libgraclus.a(ometis.c.o)
    /Users/vagrant/Data/installed/x64-osx/debug/lib/libmetis.a(ometis.c.o)
ld: 1 duplicate symbol for architecture x86_64
```

JackBoosY commented:
> x64_osx failed but I think it's not because of this PR:
> 
> ```
> duplicate symbol '_METIS_NodeND' in:
>     lib/Graclus/libgraclus.a(ometis.c.o)
>     /Users/vagrant/Data/installed/x64-osx/debug/lib/libmetis.a(ometis.c.o)
> ld: 1 duplicate symbol for architecture x86_64
> ```

Yeah, that's resolved.

vicroms commented:
The team discussed this PR and the general approach to improve our Linux dynamic experience:
For the most part we like the changes in this PR but there are some things that we would like to see changed:

* We would like this to work on more than just x64, this means finding a way to acquire patchelf for different platforms.
* We would like this to be applied to each package built in the dynamic triplet automatically, this means proposing a way to hook up processes per triplet (we have a few suggestions):
    * Find a way to do this by setting `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` as proposed by [this comment](https://github.com/microsoft/vcpkg/discussions/19127#discussioncomment-1058501).
    * Introduce a new triplet variable where we could hook post processing scripts, e.g.: `VCPKG_POST_BUILD_SCRIPTS` (name is a placeholder).
* If this is to be a port function, we would like to move it out of `scripts/cmake` into a proper scripts port.

Osyotr commented:
@vicroms
> We would like this to work on more than just x64, this means finding a way to acquire patchelf for different platforms.

There are prebuilt binaries for many architectures: https://github.com/NixOS/patchelf/releases/tag/0.14.5, but should we consider Windows->Linux cross compilation scenario? I have no idea whether it's possible to build patchelf for Windows. Anyway, it may be done in later PR.

> * Find a way to do this by setting `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` as proposed by [this comment](https://github.com/microsoft/vcpkg/discussions/19127#discussioncomment-1058501).

The comment you linked describes the way to automatically pass `-Wl,-rpath,'$ORIGIN'` flag to triplets. This would still require post-build patching because `-Wl,-rpath,'$ORIGIN' -Wl,-rpath,'/path/on/host'` will result in rpath `$ORIGIN:/path/on/host` which is undesired and would require patching ports. Also, many ports do not respect custom linker flags.
> * Introduce a new triplet variable where we could hook post processing scripts, e.g.: `VCPKG_POST_BUILD_SCRIPTS` (name is a placeholder).

This is what I was thinking about, too. So the whole `x64-linux-dynamic` triplet would look like this:
```cmake
set(VCPKG_TARGET_ARCHITECTURE x64)
set(VCPKG_CRT_LINKAGE dynamic)
set(VCPKG_LIBRARY_LINKAGE dynamic)

set(VCPKG_CMAKE_SYSTEM_NAME Linux)

set(VCPKG_POST_BUILD_FIXUP_RPATH ON)
```
Same for BSD and other platforms that use elf format.
> If this is to be a port function, we would like to move it out of `scripts/cmake` into a proper scripts port.
I would prefer to not make this function public. Would it be acceptable to move it into `scripts/fixup_rpaths.cmake`?

Also, is there any conclusion on what to do with tools (https://github.com/microsoft/vcpkg/issues/17607)? This is something I didn't think of very well in my initial implementation. Having multiple combinations of tools paths increases complexity of this PR. Main issue is debug variant of tools.

Neumann-A commented:
If a port is adding rpaths like described:

> -Wl,-rpath,'$ORIGIN' -Wl,-rpath,'/path/on/host'

this needs to be patch out any way. Ports are not allow to set rpaths, only the toolchain is. 

> Also, many ports do not respect custom linker flags.

this needs to be fixed. The toolchain file should define all flags and pass them on to the build.

Osyotr commented:
@Neumann-A you are right, it's better to build with correct rpath than to fix afterwards. This would require writing a post-build check in `vcpkg-tool` and fixing affected ports one by one. Unfortunately, I don't have enough knowledge, experience and time to do that.
I also don't see why we can't have both solutions simultaneously. Once all ports and build systems are fixed, we can drop post-build fixup.

Osyotr commented:
Also, without resolving https://github.com/microsoft/vcpkg/issues/17607 it will be hard to set correct rpath for tools at build time.

Neumann-A commented:
> correct rpath for tools at build time.

that will be only `$ORIGIN` and go the windows way. Either copy the *.so or symlink it into the folder. I think that is the only valid approach considering the circumstances.

Osyotr commented:
> Either copy the *.so or symlink it into the folder

Why? What's the difference between this and setting rpath to `$ORIGIN/../lib`?

Neumann-A commented:
> Why? What's the difference between this and setting rpath to $ORIGIN/../lib?

It doesn't require looking into the parent if there is a `/lib` folder which does not exists if it is installed into `tools/port/bin` or `tools/port/`.
The Idea of doing `$ORIGIN/../lib` is that the folder structure is such that there is a `bin` folder and besides it is a `lib` folder. I personally don't mind it but see that it requires a fixed layout which vcpkg simply does not have. As such setting it to $ORIGIN and do it the windows way seems simply more correct and easier to implement.

Osyotr commented:
> It doesn't require looking into the parent if there is a `/lib` folder which does not exists if it is installed into `tools/port/bin` or `tools/port/`.

I'm already [calculating](https://github.com/microsoft/vcpkg/pull/23035/files#diff-fba81f18c23594993760548d2a3a7d5a532f4ebcfb16baa98305121e4c29f713R39) relative path to `lib/` in this PR and release tools do have right rpath (`$ORIGIN/../../lib` or `$ORIGIN/../../../lib` etc.)
The only issues are debug tools but I guess it can be fixed by checking whether a tool is located in `debug/` subdirectory and if it's true then use `$ORIGIN/../~/../debug/lib`. I don't think vcpkg should install debug tools, though.

dg0yt commented:
> I don't think vcpkg should install debug tools, though.

With the current ad-hoc layout, debug tools also need help to find data files. For Linux, it would be much easier if debug variants were installed into an entirely separate prefix.

Neumann-A commented:
> Introduce a new triplet variable where we could hook post processing scripts,

hrhr. I have an idea: somebody tried [cmake_language()](https://cmake.org/cmake/help/latest/command/cmake_language.html#defer) to inject post build stuff ?

Osyotr commented on triplets/x64-linux.cmake:
```diff
@@ -1,6 +1,7 @@
 set(VCPKG_TARGET_ARCHITECTURE x64)
 set(VCPKG_CRT_LINKAGE dynamic)
-set(VCPKG_LIBRARY_LINKAGE static)
+set(VCPKG_LIBRARY_LINKAGE dynamic)
```
This is only to trigger CI build for dynamic triplet. Should be reverted before merge.

Osyotr commented:
Here's summary of changes since last review:
1) Introduced `VCPKG_POST_BUILD_SCRIPTS` triplet variable.
2) Fixed rpath in debug tools
3) `patchelf` is updated to support `aarch64`

vicroms commented on scripts/ports.cmake:
```diff
@@ -145,6 +145,9 @@ if(CMD STREQUAL "BUILD")
 
     include("${CURRENT_PORT_DIR}/portfile.cmake")
     if(DEFINED PORT)
+        foreach(VCPKG_POST_BUILD_SCRIPT IN LISTS VCPKG_POST_BUILD_SCRIPTS)
+            include("${VCPKG_POST_BUILD_SCRIPT}")
+        endforeach()
```
```suggestion
        if(VCPKG_FIXUP_ELF_RPATH)
            include("${SCRIPTS}/cmake/z_vcpkg_fixup_rpath.cmake")
        endif()
```

vicroms commented on triplets/community/x64-linux-dynamic.cmake:
```diff
@@ -0,0 +1,7 @@
+
+set(VCPKG_CMAKE_SYSTEM_NAME Linux)
+
+set(VCPKG_POST_BUILD_SCRIPTS "${SCRIPTS}/cmake/z_vcpkg_fixup_rpath.cmake")
```
```suggestion
set(VCPKG_FIXUP_ELF_RPATH ON)
```

vicroms reviewed (changes requested):
Hi @Osyotr 

We believe this is a really good change and looks about ready to merge, with just one last change requested: we would like to go back to enabling this with a single setting in the triplet file instead of the generic post-build scripts list (team suggests `VCPKG_FIXUP_ELF_RPATH`). 

The reason is that in the future we want to make patching `RPATH` a built-in feature of vcpkg, and if we did that, we would prefer the UI to be a single setting in the triplet that vcpkg knows how to handle specially.

We will leave the implementation of `VCPKG_POST_BUILD_SCRIPTS` for later when a stronger use case appears.

Osyotr committed:
Switch back to a single variable VCPKG_FIXUP_ELF_RPATH

Co-authored-by: Victor Romero <romerosanchezv@gmail.com>

Osyotr commented on triplets/x64-linux.cmake:
```diff
@@ -4,3 +4,4 @@ set(VCPKG_LIBRARY_LINKAGE static)
 
 set(VCPKG_CMAKE_SYSTEM_NAME Linux)
 
+set(VCPKG_FIXUP_ELF_RPATH ON)
```
I'm not sure how to reliably check for that.

vicroms replied:
Since this effectively changes the produced binaries we want it to be an explicit setting even if it is something that the user would actually want to enable every time

vicroms replied:
Confirmed that executables work with relocated `installed/` folder and when shared libraries are located side-by-side 🎉

JackBoosY commented:
How about use `LD_LIBRARY_PATH` internal?

BillyONeal commented:
> How about use `LD_LIBRARY_PATH` internal?

We generally expect that to not be under the control of the deployed software.

Jamlys commented:
> > How about use `LD_LIBRARY_PATH` internal?
> 
> We generally expect that to not be under the control of the deployed software.
@BillyONeal What do you mean the deployed software? I don't quite understand.

Jamlys commented:
> The team discussed this PR and the general approach to improve our Linux dynamic experience: For the most part we like the changes in this PR but there are some things that we would like to see changed:
> 
> * We would like this to work on more than just x64, this means finding a way to acquire patchelf for different platforms.
> * We would like this to be applied to each package built in the dynamic triplet automatically, this means proposing a way to hook up processes per triplet (we have a few suggestions):
>   
>   * Find a way to do this by setting `VCPKG_CHAINLOAD_TOOLCHAIN_FILE` as proposed by [this comment](https://github.com/microsoft/vcpkg/discussions/19127#discussioncomment-1058501).
>   * Introduce a new triplet variable where we could hook post processing scripts, e.g.: `VCPKG_POST_BUILD_SCRIPTS` (name is a placeholder).
> * If this is to be a port function, we would like to move it out of `scripts/cmake` into a proper scripts port.

As @vicroms said, personally I also hope that this solution can work on different platforms. @Osyotr Or at least this can be an optional approach, so that another alternative solution can work by providing an overlay triplet, without impacted by this one.

Osyotr commented:
> Or at least this can be an optional approach

This is already optional. But there is really no need for it to be disabled.
> personally I also hope that this solution can work on different platforms

In this PR I've added patchelf for x64 and aarch64, but this list is easily expandable. It's also available for `armv7l`, `i686`, `ppc64le` and `s390x`, but I've never heard of anyone using these platforms for development. If you are such person, please tell us more.

Jamlys commented:
> In this PR I've added patchelf for x64 and aarch64, but this list is easily expandable. It's also available for `armv7l`, `i686`, `ppc64le` and `s390x`, but I've never heard of anyone using these platforms for development. If you are such person, please tell us more.

I am recently preparing the development environment for a new cpu family member called loongarch64. Basically I cross-compile for it so far. But I am still not so sure whether it's necessary to do the developing and debugging work on it in the future, which I don't hope so.

Jamlys commented:
BTY, to support cross-compiling, xxx-linux-dynamic triplet also gotta work with copying the host dependencies into the corresponding directory. Otherwise the corresponding rpath should be an absolute path, which I don't know whether it's suitable or not. By trying not to modify the output after compilation, here I come up with another two solution as proposals, which may not be suitable but might bring other ideas.

> 1. Use `readelf` to detect rpath information of the target binary, and transform it into absolute path. Copy the dependencies into the transformed directory. All these can be done within `vcpkg_copy_tool_dependencies`.

This is not so suitable for that it may mess up the layout of installed directory.


> 2. Wrap the specified compilers into a internal bash script if it needed, within which add addition options to provide rpath, then every target using these wrapped 'compilers' will have the specified rpath properties.

This solution is also tricky. But before this PR and none of any other PR being confirmed and merged, this is the only way I can do by providing my own overlay `linux.cmake` tool chain, without modifying vcpkg locally.

dg0yt commented:
For cross-compiling you can always use a static host triplet. Cross-compiling to mingw triplets works very well this way.

RPATH must not contain absolute paths. It breaks relocatability of cached artifacts.

Jamlys commented:
> For cross-compiling you can always use a static host triplet. Cross-compiling to mingw triplets works very well this way.
> 
> RPATH must not contain absolute paths. It breaks relocatability of cached artifacts.

Got it. Thx for suggestion. : )

Osyotr committed:
Don't force fixup in x64-linux triplet yet

JackBoosY commented:
So currently the only thing preventing us from converting x64-linux-dynamic to an official triplet is our budget?

BillyONeal commented:
Yes

silverqx commented:
Great powerful PR, thx @Osyotr

vserdyuk commented:
> > Or at least this can be an optional approach
> 
> This is already optional. But there is really no need for it to be disabled.
> 
> > personally I also hope that this solution can work on different platforms
> 
> In this PR I've added patchelf for x64 and aarch64, but this list is easily expandable. It's also available for `armv7l`, `i686`, `ppc64le` and `s390x`, but I've never heard of anyone using these platforms for development. If you are such person, please tell us more.

I am using an `armv7l` host (32-bit Ubuntu on Raspberry Pi 4B). Had to fix the script (`HOST_ARCH` is checked, but not initialized) and add the appropriate download hash. See PR for details: https://github.com/microsoft/vcpkg/pull/26510
